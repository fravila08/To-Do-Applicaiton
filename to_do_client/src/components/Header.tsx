import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import React, { useEffect, useState } from "react";
import axios from "axios";
import { ITask } from "../App";

interface HeaderProps {
  allTasks: ITask[];
  setAllTasks: (allTasks: ITask[]) => void;
}

export interface ResponseCreateTask {
  itemCreated: boolean;
  id: number;
}

export const createTask = async (taskTitle: string): Promise<ResponseCreateTask> => {
  let response = await axios.post("newtask/", {
    name: taskTitle,
  });
  return response["data"];
};

export const Header: React.FC<HeaderProps> = ({ allTasks, setAllTasks }) => {
  const [isDisabled, setIsDisabled] = useState(true);
  const [newTask, setNewTask] = useState("");

  useEffect(() => {
    let evalInput = newTask.replaceAll(" ", "");
    if (evalInput.length >= 1 && evalInput != "") {
      setIsDisabled(false);
    } else {
      setIsDisabled(true);
    }
  }, [newTask]);

  const createNewTask = async (
    str: string,
    event?: React.FormEvent<HTMLFormElement>
  ) => {
    event?.preventDefault();
    let response = await createTask(str);
    if (response.itemCreated) {
      setAllTasks([
        ...allTasks,
        { id: response.id, title: str, completed: false },
      ]);
      setNewTask("");
    }
  };

  return (
    <Container style={{ marginBottom: "1vh" }}>
      <Row>
        <Col xs={4}></Col>
        <Col xs={8} className="formHolder">
          <Form onSubmit={(e) => createNewTask(newTask, e)}>
            <Form.Control
              value={newTask}
              id="createTaskInput"
              onChange={(e) => setNewTask(e.target.value)}
            />
            <Button
              variant="success"
              id="createTaskButton"
              disabled={isDisabled}
              type="submit"
            >
              Create
            </Button>
          </Form>
        </Col>
      </Row>
    </Container>
  );
};
