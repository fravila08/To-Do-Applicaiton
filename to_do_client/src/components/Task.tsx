import Container from 'react-bootstrap/Container'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Form from 'react-bootstrap/Form'
import { useEffect, useState } from 'react'
import { Tasks } from '../App'

interface IdSelectionProps{
    selected: number[];
    setSelected: (selected: number[]) => void;
    task: Tasks;
}




export const Task: React.FC<IdSelectionProps>=({setSelected, selected, task})=>{
    return(
        <Row style={{backgroundColor:"grey", border:"black 2px solid"}}>
            <Col xs={8} style={{display:'flex'}}>
                {task.Title}
            </Col>
            <Col xs={2}>
                <Form.Check type="checkbox" label="select" 
                onChange={(e)=>{e.target.checked?
                    setSelected([...selected, task.id])
                    :
                    setSelected(selected.filter(item=> item !== task.id))
                    }} 
                />
            </Col>
        </Row>
    )
}