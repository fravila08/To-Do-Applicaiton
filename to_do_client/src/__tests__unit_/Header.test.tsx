import { describe, expect, it, vi, Mocked } from "vitest";
import TestRenderer from 'react-test-renderer';
import { Header, createTask } from "../components/Header";
import axios from 'axios';

vi.mock('axios')

describe("This will confirm the correct information will be sent on `post`",()=>{
    it("will create a new task with the current content in the input bar", async()=>{
        const mockedAxios= axios as Mocked<typeof axios>;
        mockedAxios.post.mockResolvedValue({data:{itemCreated:true}})
        let name="Style this app"
        let createtask= await createTask(name)
        expect(createtask).toBeTruthy()
    })
})


describe("Will take a snapshot of the current Header compenent",()=>{
    it("will create a snapshot",()=>{
        let selectedIDs:number[]=[]
        const header= TestRenderer.create(<Header selectedIDs={selectedIDs}/>)
        expect(header).toMatchSnapshot()
    })
})