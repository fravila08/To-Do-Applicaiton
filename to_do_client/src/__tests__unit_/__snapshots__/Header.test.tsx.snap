// Vitest Snapshot v1

exports[`Will take a snapshot of the current Header compenent > will create a snapshot 1`] = `
<div
  className="container"
  style={
    {
      "marginBottom": "1vh",
    }
  }
>
  <div
    className="row"
  >
    <div
      className="col-4"
      style={
        {
          "display": "flex",
          "justifyContent": "center",
          "margin": 0,
          "padding": 0,
        }
      }
    >
      <button
        className="btn btn-danger"
        disabled={true}
        type="button"
      >
        Del
      </button>
      <button
        className="btn btn-warning"
        disabled={true}
        type="button"
      >
        Stat
      </button>
    </div>
    <div
      className="col-8"
      style={
        {
          "margin": 0,
          "padding": 0,
        }
      }
    >
      <form
        className=""
        onSubmit={[Function]}
        style={
          {
            "display": "flex",
          }
        }
      >
        <input
          className="form-control"
          onChange={[Function]}
          value=""
        />
        <button
          className="btn btn-success"
          disabled={true}
          type="submit"
        >
          Create
        </button>
      </form>
    </div>
  </div>
</div>
`;
